class-two-factor-core.php

```diff
@@ -775,8 +775,48 @@ public static function login_html( $user, $login_nonce, $redirect_to, $error_msg
 			#login form p.two-factor-prompt {
 			margin-bottom: 1em;
 			}
+			.input.authcode {
+				letter-spacing: .3em;
+			}
+			.input.authcode::placeholder {
+				opacity: 0.5;
+			}
 		</style>
-
+		<script>
+			(function() {
+				// Enforce numeric-only input for numeric inputmode elements.
+				const form = document.querySelector( '#loginform' ),
+					inputEl = document.querySelector( 'input.authcode[inputmode="numeric"]' ),
+					expectedLength = inputEl?.dataset.digits || 0;
+
+				if ( inputEl ) {
+					let spaceInserted = false;
+					inputEl.addEventListener(
+						'input',
+						function() {
+							let value = this.value.replace( /[^0-9 ]/g, '' ).trimStart();
+
+							if ( ! spaceInserted && expectedLength && value.length === Math.floor( expectedLength / 2 ) ) {
+								value += ' ';
+								spaceInserted = true;
+							} else if ( spaceInserted && ! this.value ) {
+								spaceInserted = false;
+							}
+
+							this.value = value;
+
+							// Auto-submit if it's the expected length.
```

Comment:
I like the convenience of this when it works, but I also sometimes run into a problem where I make a typo on the last digit. When that happens the page refreshes and I have to start over.

I think ideally it'd use an XHR, so that the user can pick up where they left off if there's an error. That's probably more trouble than it's worth, though.

I'm not opposed to merging this, it's probably an improvement on average. Just noting the potential problems in case anyone has a good solution.

Reply:
That's fair, I'm happy to remove this and put it into a second PR, which I think I'll do.

Reply:
Either way 👍🏻